Added initial documentation for maintainers (#34)
php/pie: docs/extension-maintainers.md

# PIE for Extension Maintainers

Adding PIE support for your extension is relatively straightforward, and the
flow is quite similar to adding a regular PHP package into Packagist.

## Add a `composer.json` to your extension

The first step to adding PIE support is adding a `composer.json` to your
extension repository. Most of the typical fields are the same, with a few
notable exceptions:

 * The `type` MUST be either `php-ext` for a PHP Module (this will be most
   extensions), or `php-ext-zend` for a Zend Extension.
 * An additional `php-ext` section MAY exist.
 * The Composer package name (i.e. the top level `name` field) for a PIE
   extension cannot share the same Composer package name as a regular PHP

>    package, even if they have different `type` fields.

Reviewer: Maybe it's worth mentioning that the package name should follow the usual convention for Composer packages (i.e. `<vendor>/<package>` instead of the "usual" `ext-<extension>`) in addition to being unique across the Composer ecosystem.

### The `php-ext` definition

#### `extension-name`

The `extension-name` SHOULD be specified, and must conform to the usual extension
name regular expression, which is defined in
[\Php\Pie\ExtensionName::VALID_PACKAGE_NAME_REGEX](../src/ExtensionName.php).
If the `extension-name` is not specified, the `extension-name` will be derived
from the Composer package name, with the vendor prefix removed. For example,
given a `composer.json` with:

```json
{
    "name": "myvendor/myextension"
}
```

The extension name would be derived as `myextension`. The `myvendor/` vendor
prefix is removed.

> The `extension-name` MAY be prefixed with `ext-` as is a convention in Composer,
> but this is optional.

Reviewer: I don't remember all of the context behind this decision, but as a best practice for new authors I think it's better to suggest people don't prefix their extension name, as `ext-` is only used to distinguish extensions from packages and libraries in an otherwise ambiguous name field. Since we're in the context of a PHP extension here, the `ext-` prefix is implied and should be omitted from the name.

Consider this:
```suggestion
The `extension-name` SHOULD NOT be prefixed with `ext-` as is a convention in Composer.
```

Instead of suggesting that users can use the prefix, this wording suggests that authors omit it, without saying that it's forbidden. I understand that leaves a grey area ("is it supported?"), but I'd much rather avoid a useless prefix.

            }
        ]
    }
}
```

Then the `pie build` or `pie install` commands may be invoked in the following
ways to achieve the desired configuration:

 * `pie install myvendor/myext`
   * This will simply invoke `./configure` without any parameters
 * `pie install myvendor/myext --enable-my-flag`
   * This will invoke `./configure --enable-my-flag`
 * `pie install myvendor/myext --with-some-lib=/path/to/somelib`
   * This will invoke `./configure --with-some-lib=/path/to/somelib`
 * `pie install myvendor/myext --enable-my-flag --with-some-lib=/path/to/somelib`
   * This will invoke `./configure --enable-my-flag --with-some-lib=/path/to/somelib`

Note that it is not possible for end users of PIE to specify configuration
options that have not been defined in your extension's `configure-options`
definition. Using the same example above `composer.json`, invoking PIE with
an invalid option, such as `pie install myvendor/myext --something-else` will
result in an error `The "--something-else" option does not exist.`.

If an end user does not specify a flag defined in the `configure-options`
definition, it will simply not be passed to `./configure`. There is no way to

> specify a default value in the `configure-options` definition.

Reviewer: Thank you for this entire chapter - that cleans up most of the questions that came up while I was adding PIE support for the MongoDB extension.

> ## Submit the extension to Packagist

Reviewer: Could you add a note that package authors also need to tag a release for the extension to be installable until support for building from commit references is added? Otherwise, this may confuse users that are used to not needing a tag to test a package.

Reviewer 2: AFAIK, you can already do this, or at least from `dev-master`: 
```
pie install xdebug/xdebug:dev-master
```

Author: Since #13 was fixed, it should be possible as mentioned. I'll check Mongo shortly, and if it's not working try to figure out why separately

> # PIE for Extension Maintainers

Author: Need to add a section about Windows and how DLLs need to be published.
